SPIFlashLogger.device.lib.nut

```diff
@@ -0,0 +1,956 @@
+        function isValid() {
+            return _isValid;
+        }
+
+        //
+        // Return object address
+        //
+        function getAddress() {
+            return _addr;
+        }
+
+        //
+        // De-serialize object data and return payload
+        //
+        function getPayload() {
+            if (!_isValid)
+                return null;
+
+            if (_payload != null)
+                return _payload;
+
+            local pos = _addr;
+            local serialised = blob(SPIFLASHLOGGER_OBJECT_HDR_SIZE + _len);
+            local leftInObject;
+            _logger._enable();
+            // while there is more object left, read as much as we can from each sector into `serialised`
+            while (leftInObject = serialised.len() - serialised.tell()) {
+                // Decide what to read
+                local sectorStart = pos - (pos % SPIFLASHLOGGER_SECTOR_SIZE);
+                local sectorEnd = sectorStart + SPIFLASHLOGGER_SECTOR_SIZE;
+                local leftInSector = sectorEnd - pos;
+
+                // Read it
+                local read;
+                local left = leftInObject < leftInSector ? leftInObject : leftInSector;
+                read = _flash.read(pos, left);
+                assert(read.len() == left);
+
+                serialised.writeblob(read);
+
+                // Update remaining and position
+                leftInObject -= read.len();
+
+                pos += read.len();
+                assert(pos <= sectorEnd);
+
+                if (pos == _logger._end) pos = _logger._start + SPIFLASHLOGGER_SECTOR_METADATA_SIZE;
+                else if (pos == sectorEnd) pos += SPIFLASHLOGGER_SECTOR_METADATA_SIZE;
+
+            }
+            _logger._disable();
+
+            // Try to deserialize the object
+            _payload = null;
+            try {
+                _payload = _logger._serializer.deserialize(serialised, SPIFLASHLOGGER_OBJECT_MARKER);
+            } catch (e) {
+                //server.error(format("Exception reading logger object address 0x%04x with length %d: %s", _addr, serialised.len(), e));
+            }
+            return _payload;
+        }
+
+        //
+        // Erase object startcode
+        // and make object invisible for the next iterations
+        function erase() {
+            _logger._enable();
+            local clear = blob(SPIFLASHLOGGER_OBJECT_MARKER_SIZE);
+            local res = _flash.write(_addr, clear, SPIFLASH_POSTVERIFY);
+            _logger._disable();
+
+            if (res != 0) {
+                server.error("Clearing object marker failed.");
+                return false;
+            }
+
+            return true;
+        }
+    }
+
+    // Logger sector class
+    // This class is responsible for the logger sector's metadata
+    // reading and update;
+    // This class write all sector write operations
+    // and update sectors metadata accordingly
+    LoggerSector = class {
```

Comment:
as it's supposed to be a private field, please use `_` prefix to mark that it's private